src/versionbits.cpp

```diff
@@ -49,20 +57,17 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
         pindexPrev = vToCompute.back();
         vToCompute.pop_back();
 
+        // We track state by previous-block, so the height we should be comparing is +1
+        const int64_t height = pindexPrev->nHeight + 1;
+
         switch (state) {
             case ThresholdState::DEFINED: {
-                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
-                    stateNext = ThresholdState::FAILED;
-                } else if (pindexPrev->GetMedianTimePast() >= nTimeStart) {
+                if (height >= height_start) {
                     stateNext = ThresholdState::STARTED;
                 }
                 break;
             }
             case ThresholdState::STARTED: {
-                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
-                    stateNext = ThresholdState::FAILED;
-                    break;
-                }
                 // We need to count
                 const CBlockIndex* pindexCount = pindexPrev;
```

Comment:
By starting here, are we then ignoring if the latest block is signaling?

context:
```
// We track state by previous-block, so the height we should be comparing is +1
const int64_t height = pindexPrev->nHeight + 1;
```

Reply:
Also for my sanity, is the reason the height check is moved from the beginning to the end because we are using `pindexPrev->nHeight + 1`

Reply:
Yes. Determining whether the threshold is reached is done at the beginning of each signaling period, and determines the state for the entire signaling period. This is checking for whether the threshold is reached in the previous signaling period to determine whether the current signaling period (starting at block `height`) is locked_in. The current block (at `height`) is not part of the previous signaling period, so it is not counted.

Reply:
Consider a block B just after a retarget period boundary, ie `B->nHeight % 2016 == 0`, and `pindexPrev = B->pprev`. `pindexPrev` was in state `STARTED`. What state should `B` be in? It depends on the signalling of the prior 2016 blocks -- so we test `pindexPrev->nVersion`, `pindexPrev->pprev->nVersion`, etc. If enough signal, we switch to `LOCKED_IN`. If they don't, we check `B->nHeight` (which is `B->prev->nHeight + 1` aka `pindexPrev->nHeight + 1`) to see if it's reached the timeout, and if so switch to `FAILED`.

The reason the height check is moved to the end is so that every retarget period that begins in `STARTED` is allowed to transition to `LOCKED_IN` -- previously, if the final period passed the timeout MTP midway through, it would be too late and would unconditionally transition to `FAILED`.

Reply:
> Also for my sanity, is the reason the height check is moved from the beginning to the end of this case because we are using `pindexPrev->nHeight + 1`

No. The height check is moved because BIP 8 specifies that the threshold being reached takes precedence over the timeoutheight being reached. This is because determining whether the threshold is reached would be done at `timeoutheight`. So if the height check came first, the last signaling period would always fail. Instead, we want to succeed if the threshold is reached in the  last signaling period.